pkg/handlers/v1/books.go

```diff
@@ -2,56 +2,121 @@ package v1
+	"book_service/pkg/models/common/res"
 	"book_service/pkg/query"
 	"book_service/pkg/utils"
-	"fmt"
+	"net/http"
+
 	"github.com/gin-gonic/gin"
 	"github.com/google/uuid"
 	"github.com/jinzhu/copier"
 	"github.com/sirupsen/logrus"
-	"net/http"
 )
 
 var booksIndex, _ = utils.GetEnvVar[string]("BOOKS_INDEX", "books")
 
 func GetBookById(c *gin.Context) {
-	payload, _ := c.Get("validated")
-	bookId, _ := payload.(req.GetBook)
+	bookReq := utils.GetValidatedPayload[req.GetBook](c)
 
-	// TODO: it should be in the validation level with extra validators and transformations
-	_, parseErr := uuid.Parse(bookId.ID)
-	if parseErr != nil {
-		c.JSON(http.StatusBadRequest, gin.H{"error": parseErr})
-		return
-	}
-
-	esQuery := query.NewQueryBuilder().ID(bookId.ID).Build()
-	fmt.Println(esQuery)
-	hits, err := clients.DoSearch(c, booksIndex, esQuery)
+	esQuery := query.NewQueryBuilder().ID(bookReq.ID).Build()
+	hits, _, err := clients.DoSearch(c, booksIndex, esQuery, 1, 0)
 	if err != nil {
-		c.JSON(http.StatusInternalServerError, gin.H{"message": err})
+		logrus.Errorf("Error searching for book with ID %s: %v", bookReq.ID, err)
+		c.JSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
 		return
 	}
 	if len(hits) == 0 {
+		logrus.Infof("Book with ID %s not found", bookReq.ID)
 		c.JSON(http.StatusNotFound, gin.H{"message": "Book not found"})
 		return
 	}
 
-	c.JSON(200, hits[0])
+	logrus.Infof("Book with ID %s retrieved successfully", bookReq.ID)
+	c.JSON(http.StatusOK, hits[0])
 }
 
 func CreateBook(c *gin.Context) {
-	payload, _ := c.Get("validated")
-	bodyBookReq, _ := payload.(req.AddBook)
-	book := common.Book{ID: uuid.New(), PublishDate: bodyBookReq.PublishDate.Format("2006-01-02")}
+	bodyBookReq := utils.GetValidatedPayload[req.AddBook](c)
+	book := common.Book{
+		ID:          uuid.New(),
+		PublishDate: bodyBookReq.PublishDate.Format("2006-01-02"),
+	}
+
+	if err := copier.Copy(&book, &bodyBookReq); err != nil {
+		logrus.Errorf("Error copying AddBook payload to Book: %v", err)
+		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to map request data"})
+		return
+	}
+
+	clients.EnqueueIndexTask(c, booksIndex, book.ID.String(), book, _const.CreateIndex)
+	logrus.Infof("Book with ID %s queued for creation successfully", book.ID)
+	c.JSON(http.StatusAccepted, res.AddBook{ID: book.ID})
+}
+
+func UpdateBook(c *gin.Context) {
+	bodyBookReq := utils.GetValidatedPayload[req.UpdateBook](c)
+	titleUpdate := common.TitleUpdate{}
+	if err := copier.Copy(&titleUpdate, &bodyBookReq); err != nil {
+		logrus.Errorf("Error copying UpdateBook payload to TitleUpdate: %v", err)
+		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to map request data"})
+		return
+	}
+
+	clients.EnqueueIndexTask(c, booksIndex, bodyBookReq.ID, titleUpdate, _const.UpdateIndex)
+	logrus.Infof("Book with ID %s queued for update successfully", bodyBookReq.ID)
+	c.JSON(http.StatusAccepted, res.UpdateBook{ID: uuid.MustParse(bodyBookReq.ID)})
```

Comment:
I read that `uuid.MustParse` panics if it fails - is this the desired behavior?

Reply:
Because we did the validation already in the middleware, it's ok to to call `uuid.MustParse`. But it's actually really tricky, because functions should be atomic but to what level?